* From the summary statistic headers, can you tell what reference genome versions are used for each study?

* What is the header of each file? What does the -n 1 parameter do in head? In terminal:

* How many variants are in each of the files?

* What imputation panel was used for GLGC?

* What type of meta-analysis did you run (fixed or random effects? sample size or inverse variance based?) What is the difference?

* Did you use genomic control? In what situations is it useful to use genomic control?

* What does it mean to set the minimum weight to 10,000?

* What is the difference between "ANALYZE" and "ANALYZE HETEROGENEITY"?

* How might you create the config file if your summary statistics files had different header labels?

* What was the smallest p-value and how many markers was the meta-analysis completed for?

* Do you think we will we use the same genome-wide significance threshold (5xE-8) for the meta-analysis as we used for the GWAS? Why or why not?

* How many genome-wide significant results are there now?

* How does the inflation appear to you?

* What is the lambda value for the smallest minor allele frequency (MAF) bin?
